### DISCUSSION TITLE
how to speed up the "resolve image config" and "load metadata"?

### DISCUSSION THREAD

**Discussion by llwwbb** [docker/buildx#2110] [Q&A]
I'm using the gitlab shell executor to run
```
docker buildx build -t xxx .
```
and my dockerfile looks like
```dockerfile
# syntax=docker/dockerfile:1
FROM golang:1.21-alpine3.18 as builder
...
FROM alpine:3.18
COPY --from=builder main .
```
The execution result is like this
```
#0 building with "default" instance using docker driver
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 1.18kB done
#1 DONE 0.0s
#2 [internal] load .dockerignore
#2 transferring context: 116B done
#2 DONE 0.0s
#3 resolve image config for docker.io/docker/dockerfile:1
#3 DONE 15.3s
#4 docker-image://docker.io/docker/dockerfile:1@sha256:42399d4635eddd7a9b8a24be879d2f9a930d0ed040a61324cfdf59ef1357b3b2
#4 CACHED
#5 [internal] load metadata for docker.io/library/golang:1.21-alpine3.18
#5 DONE 0.0s
#6 [internal] load metadata for docker.io/library/alpine:3.18
#6 DONE 16.5s
#7 [builder 1/7] FROM docker.io/library/golang:1.21-alpine3.18
#7 DONE 0.0s
#8 [stage-1 1/4] FROM docker.io/library/alpine:3.18@sha256:eece025e432126ce23f223450a0326fbebde39cdf496a85d8c016293fc851978
#8 DONE 0.0s
#9 [internal] load build context
#9 transferring context: 12.07kB done
#9 DONE 0.0s
```
Can anyone help me why "resolve image config for docker.io/docker/dockerfile:1" and "load metadata for docker.io/library/alpine:3.18" take so long? I know China's international network connections are very slow, but these images should be cached locally and don't need to request the network every time, right? What can I do to speed things up?

**Comment by sanmai-NL** [comment]
It's one of those [old Docker Inc. tooling features](https://docs.docker.com/build/dockerfile/frontend/) with little value, nowadays. `Dockerfile` syntaxes aren't so dynamic and incompatible tooling is a non-problem.

Using a fixed SemVer `# syntax=docker/dockerfile:1.6.0` cuts the delay from around 7 seconds to 3 here (first build, subsequent ones hopefully use a cached version of this image).

**Reply by silencej** [reply]
Yes, specifying the exact semver can trigger the cache.

**Comment by fenollp** [comment]
Use full locked image identifier:
```Dockerfile
# syntax=docker.io/docker/dockerfile:1@sha256:ac85f380a63b13dfcefa89046420e1781752bab202122f8f50032edf31be0021
FROM docker.io/library/golang:1.21-alpine3.18@sha256:9390a996e9f957842f07dff1e9661776702575dd888084e72d86eaa382ad56e3 as builder
...
```
this way the resolver has enough info to do an image cache check. If then there are still network calls (on a cache hit) then there's a bug.

**Reply by liubog2008** [reply]
It seems not always work? Sometimes it takes 9s and sometimes it takes 0s
